python/build/lib/pcc/dataframe.py

```diff
@@ -1,149 +1,835 @@
+                self.object_map.setdefault(
+                    othertpname, 
+                    RecursiveDictionary())[new_obj.__primarykey__] = new_obj
+
+        for othertp in old_memberships:
+            othertpname = othertp.__realname__
+            for id in set(old_memberships[othertp]).difference(set(id_map[othertp])):
+                if self.start_recording or othertpname in self.tp_to_attached_df:
+                    self.add_to_record_cache(Event.Delete, othertpname, id, None)
+                del self.object_map[othertpname][id]
+
+        return new_objs_map
+
+    def __adjust_pcc(self, objs, groupname, original_changes = None):
+        if not self.calculate_pcc:
+            return
+        can_be_created = []
+        for othertp in self.group_to_members[groupname]:
+            other_cats = self.categories[othertp.__realname__]
+            if ((not self.__is_impure(othertp, other_cats))
+                and len(set([ObjectType.Projection,
+                             ObjectType.Subset,
+                             ObjectType.Union]).intersection(
+                                 other_cats)) > 0):
+                can_be_created.append(othertp)
+
+        old_memberships = {}
+        for othertp in can_be_created:
+            othertpname = othertp.__realname__
+            old_set = old_memberships.setdefault(othertp, set())
+            if (othertpname in self.known_objects):
+                old_set.update(set([id for id in self.known_objects[othertpname] if id in objs]))
+            
+        new_objs_map = self.__record_pcc_changes(self.__calculate_pcc(can_be_created, {groupname: objs}, None),
+                                                 old_memberships,
+                                                 {groupname: original_changes} if original_changes else None)
+        for othertp, new_objs in new_objs_map.items():
+            for new_obj in new_objs:
+                self.object_map.setdefault(
+                    othertp.__realname__, 
+                    RecursiveDictionary())[new_obj.__primarykey__] = new_obj
+
+    def __make_pcc(self, pcctype, relevant_objs, param_map, params, hasSingleton = False, hasCollection = False):
+        universe = []
+        param_list = []
+        for tp in pcctype.__ENTANGLED_TYPES__:
+            universe.append(relevant_objs[tp])
+
+        # Creating random rule for parameters, all Collections first, then singleton objects
+        # Why? Random, couldnt think of a better way
+
+        if hasCollection:
+            for tp in pcctype.__parameter_types__[ParameterMode.Collection]:
+                param_list.append(param_map[tp])
+        if hasSingleton:
+            param_list.extend(params)
+
+        try:
+            pcc_objects = create(pcctype, *universe, params = param_list)
+        except TypeError, e:
+            print ("Exception in __make_pcc: " + e.message)
```

Comment:
Printing an exception like this will not give us a stack trace, which is annoying to fix. Try this: http://stackoverflow.com/questions/3702675/how-to-print-the-full-traceback-without-halting-the-program
There may be other places you are printing exceptions like this, so fixing those would be nice too